demes/demes.py

```diff
@@ -1688,6 +1696,51 @@ def _add_pulse(self, *, source, dest, proportion, time) -> Pulse:
         self.pulses.append(pulse)
         return pulse
 
+    def _migration_matrices(self):
+        """
+        Return a list of migration matrices.
+        """
+        uniq_times = set(migration.start_time for migration in self.migrations)
+        uniq_times.update(migration.end_time for migration in self.migrations)
+        end_times = sorted(uniq_times, reverse=True)[1:]
+        n = len(self.demes)
+        mm_list = [[[0] * n for _ in range(n)] for _ in range(len(end_times))]
+        deme_id = {deme.name: j for j, deme in enumerate(self.demes)}
+        for migration in self.migrations:
+            start_time = math.inf
+            for k, end_time in enumerate(end_times):
+                if start_time <= migration.end_time:
+                    break
+                if end_time < migration.start_time:
+                    if isinstance(migration, AsymmetricMigration):
+                        source_id = deme_id[migration.source]
+                        dest_id = deme_id[migration.dest]
+                        mm_list[k][dest_id][source_id] = migration.rate
+                    else:
+                        for source, dest in itertools.permutations(migration.demes, 2):
+                            source_id = deme_id[source]
+                            dest_id = deme_id[dest]
+                            mm_list[k][dest_id][source_id] = migration.rate
+                start_time = end_time
+        return mm_list, end_times
+
+    def _check_migration_rates(self):
+        """
+        Check that the sum of migration ingress rates doesn't exceed 1 for any
+        deme in any interval of time.
+        """
+        start_time = math.inf
+        mm_list, end_times = self._migration_matrices()
+        for migration_matrix, end_time in zip(mm_list, end_times):
+            for j, row in enumerate(migration_matrix):
+                if sum(row) > 1:
```

Comment:
Maybe add a ``and not math.isclose(row_sum, 1)`` to allow for the case when the sum is 1.0000000001 (which happens).